### PR TITLE
feat: add aliyun captcha

### PR DESCRIPTION
**Pull request by Resulte**
https://user-images.githubusercontent.com/34178764/176194718-0256e1c4-cb83-49fa-bc53-a62170058de9.mov

### CODE REVIEW

File: util/crypto.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+package util
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/base64"
+)
+
+func HMACSHA1(keyStr, value string) string {
```
Comment (nomeguy) [inline util/crypto.go:23]: Use name like `GetHmacSha1`

File: web/src/common/CaptchaPreview.js [hunk 2]
```diff
@@ -100,24 +110,40 @@ export const CaptchaPreview = ({ provider, providerName, clientSecret, captchaTy
...
 
   return (
     <React.Fragment>
-      <Button style={{ fontSize: 14 }} type={"primary"} onClick={clickPreview} disabled={captchaType !== "Default" && (!clientId || !clientSecret)}>
+      <Button
+        style={{ fontSize: 14 }}
+        type={"primary"}
+        onClick={clickPreview}
+        disabled={
+          (captchaType !== "Default" && (!clientId || !clientSecret)) ||
```
Comment (nomeguy) [inline web/src/common/CaptchaPreview.js:142]: this logic is too complicated, define `disabled` with multi-line if-else outside of the return.

File: web/src/Setting.js [hunk 3]
```diff
@@ -118,6 +118,10 @@ export const OtherProviderInfo = {
     "hCaptcha": {
       logo: `${StaticBaseUrl}/img/social_hcaptcha.png`,
       url: "https://www.hcaptcha.com",
+    },
+    "Aliyun Captcha": {
+      logo: `${StaticBaseUrl}/img/social_aliyun.png`,
```
Comment (nomeguy) [inline web/src/Setting.js:123]: Is this logo there?

File: web/src/ProviderEditPage.js [hunk 4]
```diff
@@ -377,6 +401,32 @@ class ProviderEditPage extends React.Component {
             </React.Fragment>
           )
         }
+        {
+          this.state.provider.type !== "Aliyun Captcha" ? null : ( 
+            <>
+              <Row style={{marginTop: '20px'}} >
```
Comment (nomeguy) [inline web/src/ProviderEditPage.js:407]: clientId2 and clientSecret2 are already in UI?

**Review by nomeguy — CHANGES_REQUESTED** [review]

File: web/src/common/CountDownInput.js [hunk 5]
```diff
@@ -79,13 +82,14 @@ export const CountDownInput = (props) => {
...
         setCheckType("Default");
         setVisible(true);
-      } else if (res.type === "reCAPTCHA" || res.type === "hCaptcha") {
+      } else {
         setCheckType(res.type);
         setClientId(res.clientId);
         setCheckId(res.clientSecret);
         setVisible(true);
-      } else {
-        Setting.showMessage("error", i18next.t("signup:Unknown Check Type"));
```
Comment (seriouszyx) [inline web/src/common/CountDownInput.js:88]: if `Setting` is not used, plz remove import codes `import * as Setting from "../Setting";` to fix the warning

image

**Review by nomeguy — APPROVED** [review]

**Pull request merged**